src/storage/src/sink/kafka.rs

```diff
@@ -1343,26 +1344,12 @@ fn encode_collection<G: Scope>(
             *capset = CapabilitySet::new();
 
             let mut row_buf = Row::default();
+            let mut datums = DatumVec::new();
 
             while let Some(event) = input.next().await {
                 if let Event::Data(cap, rows) = event {
                     for ((key, value), time, diff) in rows {
-
-                        let headers = match connection.headers_index {
-                            Some(i) => {
-                                // Only the upsert envelope makes it unambiguous
-                                // as to whether to use `before` or `after` to
-                                // compute the headers. The rule is simple: use
-                                // `after` if it exists (insertions and
-                                // updates), otherwise fall back to `before`
-                                // (deletions).
-                                assert!(envelope == SinkEnvelope::Upsert);
-                                let row = value.after.as_ref().or(value.before.as_ref()).expect("one of before or after must be set");
-                                encode_headers(row.iter().nth(i).unwrap())
-                            }
-                            _ => vec![],
-                        };
-                        let (hash, key) = match key {
+                        let (mut hash, key) = match key {
```

Comment:
nit: we are hashing the full key here either way, even if the `hash` is later overwritten using the partition_by method. Could we move this after the block below such that we only hash the key if it's not already set by the `partition_by` expression?

Reply:
Ah good point. Futzed with the `Encode` trait to make this possible. We now only hash the key if there isn't already a hash.